FEATURE: upload an avatar option for uploading avatars with selectable avatars

featheredtoast opened the pull request:
Allow staff or users above a trust level to upload avatars even when the site
has selectable avatars enabled.

Everyone can still pick from the list of avatars. The option to upload is shown
below the selectable avatar list.

refactored boolean site setting into an enum with the following values:
none: No selectable avatars enabled (default)
restrict_tl1: show selectable avatars, but allow tl2+ and staff to upload custom avatars
restrict_tl2: show selectable avatars, but allow tl3+ and staff to upload custom avatars
restrict_tl3: show selectable avatars, but allow tl4 and staff to upload custom avatars
restrict_nonstaff: show selectable avatars, but allow staff to upload custom
avatars.
restrict_all: old "true" setting, do not allow anyone to upload custom avatars.
All users pick from the selectable avatar list.


![Screenshot from 2022-02-09 12-38-12](https://user-images.githubusercontent.com/1322534/153285961-e0f3c31c-bce1-42ba-a863-b091b47be22d.png)

ZogStriP commented:
> restrict_tl1: show selectable avatars, but allow tl2+ and staff to upload custom avatars
> restrict_tl2: show selectable avatars, but allow tl3+ and staff to upload custom avatars
> restrict_tl3: show selectable avatars, but allow tl4 and staff to upload custom avatars
> restrict_nonstaff: show selectable avatars, but allow staff to upload custom
> avatars.
> restrict_all: old "true" setting, do not allow anyone to upload custom avatars.
> All users pick from the selectable avatar list.

Those names could be quite confusing 🤔 They are "negatives". Could we rename them to "positives" instead? A bit like we prefer using `if` instead of `unless`.

featheredtoast commented:
yeah, names are up for changing, I settled on restrict_* originally to follow the "restrict_all" wording, though I suppose I could allow_* for everything, and have restrict_all be the special value could work.

Additionally, you have better suggestions for what to call "restrict_all" and "none" here, I'm all for changing those to be more in line with allow wording.

ZogStriP commented on app/assets/javascripts/discourse/app/controllers/avatar-selector.js:
```diff
@@ -21,11 +21,40 @@ export default Controller.extend(ModalFunctionality, {
+
+  @discourseComputed("siteSettings.selectable_avatars_enabled")
+  showAvatarUploader(selectableAvatars) {
+    if (!selectableAvatars || selectableAvatars === "none") {
+      return true;
+    }
+    if (selectableAvatars === "restrict_all") {
+      return false;
+    }
+    if (selectableAvatars?.startsWith("restrict_tl")) {
+      const restrictedTl = parseInt(
+        selectableAvatars.replace("restrict_tl", ""),
+        10
+      );
+      return (
+        this.user.admin ||
+        this.user.moderator ||
+        this.user.trust_level > restrictedTl
+      );
+    }
+    if (selectableAvatars === "restrict_nonstaff") {
+      return this.user.admin || this.user.moderator;
+    }
```
I think this is one of the times a `switch` is clearer as it lists all the possible values of the enum.

```suggestion
    switch (selectableAvatars) {
      case "restrict_all":
        return false;
      case "restrict_tl1":
      case "restrict_tl2":
      case "restrict_tl3":
        const restrictedTl = parseInt(
          selectableAvatars.replace("restrict_tl", ""),
          10
        );
        return (
          this.user.admin ||
          this.user.moderator ||
          this.user.trust_level > restrictedTl
        );
      case "restrict_nonstaff":
        return this.user.admin || this.user.moderator;
      case "none":
      default:
        return true;
    }
```

ZogStriP commented on app/models/user.rb:
```diff
@@ -1261,7 +1261,7 @@ def create_user_profile
 
   def set_random_avatar
-    if SiteSetting.selectable_avatars_enabled?
+    unless SiteSetting.selectable_avatars_enabled == "none"
```
Writing the condition using the "positive" style feels easier to read

```suggestion
    if SiteSetting.selectable_avatars_enabled != "none"
```

ZogStriP commented:
Maybe we should rename the site setting (and validator) to `selectable_avatars` and the choices would be 

- disabled (default)
- everyone (aka restrict_all, aka old "true" value)
- up_to_tl2
- up_to_tl3
- up_to_tl4
- non_staff

?

featheredtoast commented:
hmm, under_tl* perhaps as an alternative to up_to_tl*? The logic will have to change slightly to account for "disallow below this" vs "disallow up to this" but it's another possibility if that makes it more clear.

I do like the rename to `selectable_avatars_for`, I'll go in that direction :+1:

featheredtoast commented:
Updated the name:
selectable_avatar_restriction
      - disabled
      - no_one
      - under_tl1
      - under_tl2
      - under_tl3
      - under_tl4
      - non_staff
      - everyone
Added "no_one" to cover the use cases seen on meta, aka: have selectable avatars, but want them to be completely optional, and allow everyone, incluing tl0 to upload their own.

Alternatively we couuuuld also have it so selectable avatars show up if uploaded regardless of setting, and remove `no_one` from this list. Lots of ways to skin the cat and I'm in too deep to know what is the most intuitive.

featheredtoast commented:
Re-updated again, this time more towards the "allow" wording:

I propose naming the site setting `selectable_avatars_mode` with a description of:

> Allow users to select an avatar from the selectable_avatars list and limit custom avatar uploads to the selected trust level.

with the following options:
* disabled
* everyone
* tl1
* tl2
* tl3
* tl4
* staff
* no_one

After thinking about it, this is the clearest wording I can come up with while also having positive/permissive functionality.

ZogStriP commented:
I very much prefer that 👍

ZogStriP commented on app/assets/javascripts/discourse/app/controllers/avatar-selector.js:
```diff
@@ -17,15 +17,43 @@ export default Controller.extend(ModalFunctionality, {
   },
 
+  @discourseComputed("siteSettings.selectable_avatars_mode")
+  showSelectableAvatars(enabled) {
```
Ditto^^ `enabled` -> `mode`

ZogStriP commented on app/assets/javascripts/discourse/app/controllers/avatar-selector.js:
```diff
@@ -17,15 +17,43 @@ export default Controller.extend(ModalFunctionality, {
+  },
+
+  @discourseComputed("siteSettings.selectable_avatars_mode")
+  showAvatarUploader(selectableAvatars) {
```
```suggestion
  showAvatarUploader(mode) {
```

ZogStriP commented on db/post_migrate/20220202223955_migrate_selectable_avatars_enabled.rb:
```diff
@@ -0,0 +1,38 @@
+    execute <<~SQL
+      UPDATE site_settings AS s
+        SET value =
+          CASE WHEN t.value = 'everyone' THEN 't'
+          WHEN t.value = 'no_one' THEN 't'
+          WHEN t.value = 'staff' THEN 't'
+          WHEN t.value = 'tl1' THEN 't'
+          WHEN t.value = 'tl2' THEN 't'
+          WHEN t.value = 'tl3' THEN 't'
+          WHEN t.value = 'tl3' THEN 't'
+          WHEN t.value = 'tl4' THEN 't'
+          ELSE 'f'
+          END,
```
I wonder if it's possible to use `IN` here instead? Something like

```suggestion
          CASE WHEN t.value IN ('everyone', 'no_one', 'staff', 'tl1', ...) THEN 't'
          ELSE 'f'
          END,
```

featheredtoast replied:
Oh yes that reads much better! Updated.

featheredtoast commented:
@ZogStriP all suggestions committed :heavy_check_mark: let me know how it looks!

ZogStriP reviewed (approved):
LGTM 👏
